File: backend/models/blocks.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime


class BlockNode(BaseModel):
    id: str
    type: str
    position: Dict[str, float]
    data: Dict[str, Any]


class BlockEdge(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None
    style: Optional[Dict[str, Any]] = None


class BlockProgram(BaseModel):
    id: Optional[str] = None
    project_id: str
    name: str
    nodes: List[BlockNode]
    edges: List[BlockEdge]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ArmState(BaseModel):
    position: Dict[str, float] = Field(default_factory=lambda: {"x": 0, "y": 0, "z": 0})
    joints: List[float] = Field(default_factory=lambda: [0, 0, 0, 0, 0, 0])
    is_moving: bool = False

File: frontend/lib/api/blocks.ts
import { apiClient } from './client';

export interface BlockNode {
  id: string;
  type: string;
  position: { x: number; y: number };
  data: Record<string, any>;
}

export interface BlockEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  style?: Record<string, any>;
}

export interface BlockProgram {
  id?: string;
  project_id: string;
  name: string;
  nodes: BlockNode[];
  edges: BlockEdge[];
  created_at?: string;
  updated_at?: string;
}

export interface ArmState {
  position: { x: number; y: number; z: number };
  joints: number[];
  is_moving: boolean;
}

export const blocksApi = {
  // Block Programs
  createProgram: (program: Omit<BlockProgram, 'id' | 'created_at' | 'updated_at'>) =>
    apiClient.post<BlockProgram>('/blocks/programs', program),

  getPrograms: (projectId?: string) =>
    apiClient.get<BlockProgram[]>(`/blocks/programs${projectId ? `?project_id=${projectId}` : ''}`),

  getProgram: (programId: string) =>
    apiClient.get<BlockProgram>(`/blocks/programs/${programId}`),

  updateProgram: (programId: string, updates: Partial<BlockProgram>) =>
    apiClient.put<BlockProgram>(`/blocks/programs/${programId}`, updates),

  deleteProgram: (programId: string) =>
    apiClient.delete(`/blocks/programs/${programId}`),

  // Arm Control
  getArmState: () =>
    apiClient.get<ArmState>('/blocks/arm/state'),

  moveArmPosition: (x: number, y: number, z: number) =>
    apiClient.post(`/blocks/arm/move-position?x=${x}&y=${y}&z=${z}`),

  moveArmJoint: (joint: number, angle: number) =>
    apiClient.post(`/blocks/arm/move-joint?joint=${joint}&angle=${angle}`),

  resetArm: () =>
    apiClient.post('/blocks/arm/reset'),
};
